File: src/crossbeam/channel/examples/matching.go
// Source:
//   - https://web.archive.org/web/20171209034309/https://www.nada.kth.se/~snilsson/concurrency
//   - http://www.nada.kth.se/~snilsson/concurrency/src/matching.go
//
// Copyright & License:
//   - Stefan Nilsson
//   - Creative Commons Attribution 3.0 Unported License
//   - https://creativecommons.org/licenses/by/3.0/
//

package main

import (
	"fmt"
	"sync"
)

func main() {
	people := []string{"Anna", "Bob", "Cody", "Dave", "Eva"}
	match := make(chan string, 1) // Make room for one unmatched send.

	wg := new(sync.WaitGroup)
	for _, name := range people {
		wg.Add(1)
		go Seek(name, match, wg)
	}

	wg.Wait()

	select {
	case name := <-match:
		fmt.Printf("No one received %s’s message.\n", name)
	default:
		// There was no pending send operation.
	}
}

// Seek either sends or receives, whichever possible, a name on the match
// channel and notifies the wait group when done.
func Seek(name string, match chan string, wg *sync.WaitGroup) {
	select {
	case peer := <-match:
		fmt.Printf("%s received a message from %s.\n", name, peer)
	case match <- name:
		// Wait for someone to receive my message.
	}
	wg.Done()
}

File: src/crossbeam/channel/examples/matching.rs
use crossbeam_channel::{bounded, select};
use std::thread;

fn main() {
    let people = vec!["Alice", "Bob", "Cody", "Dave", "Eva"];
    let (s, r) = bounded(1);

    // Either send my name into the channel or receive someone else's, whatever happens first.
    let seek = |name, s: crossbeam_channel::Sender<&str>, r: crossbeam_channel::Receiver<&str>| {
        select! {
            recv(r) -> peer => println!("{} received a message from {}.", name, peer.unwrap()),
            send(s, name) -> _ => {}, // Wait for someone to receive my message.
        }
    };

    thread::scope(|scope| {
        for name in people {
            let (s, r) = (s.clone(), r.clone());
            scope.spawn(move || seek(name, s, r));
        }
    });

    if let Ok(name) = r.try_recv() {
        println!("No one received {}'s message.", name);
    }
}
